src/seraphis_mocks/scan_context_async_mock.h

```diff
@@ -0,0 +1,229 @@
+#include <vector>
+
+//forward declarations
+
+
+namespace sp
+{
+namespace scanning
+{
+namespace mocks
+{
+//-------------------------------------------------------------------------------------------------------------------
+struct ChunkRequest final
+{
+    std::uint64_t start_index;
+    std::uint64_t requested_chunk_size;
+};
+//-------------------------------------------------------------------------------------------------------------------
+struct PendingChunk final
+{
+    ChunkRequest chunk_request;
+    sp::scanning::PendingChunkContext pending_context;
+    sp::scanning::PendingChunkData pending_data;
+};
+
+static inline bool operator<(const PendingChunk &pending_chunk1, const PendingChunk &pending_chunk2)
+{
+    return pending_chunk1.chunk_request.start_index < pending_chunk2.chunk_request.start_index;
+}
+//-------------------------------------------------------------------------------------------------------------------
+struct AsyncScanContextLegacyConfig final
+{
+    /// max number of chunks that will be requested at once in a queue
+    const std::uint64_t pending_chunk_queue_size{10};
+    /// maximum number of times to retry fetching blocks from daemon on failure
+    std::uint64_t max_get_blocks_attempts{3};
+    /// whether or not user trusts the daemon's results
+    bool trusted_daemon{false};
+};
+//-------------------------------------------------------------------------------------------------------------------
+////
+// WARNING: if the chunk size increment exceeds the max chunk size obtainable from the raw chunk data source, then
+//          this will be less efficient because it will need to 'gap fill' continuously. To maximize efficiency,
+//          either make sure the scanner is pointing to a daemon that supports the max_block_count req param,
+//          or use a pending_chunk_queue_size of 1 and a multithreaded enote finding context.
+///
+class AsyncScanContextLegacy final : public ScanContextLedger
+{
+public:
+    AsyncScanContextLegacy(const AsyncScanContextLegacyConfig &config,
+            sp::EnoteFindingContextLegacy &enote_finding_context,
+            async::Threadpool &threadpool,
+            const std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request&, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response&)> &_rpc_get_blocks) :
+        m_config{config},
+        m_enote_finding_context{enote_finding_context},
+        m_scanner_ready{false},
+        m_threadpool{threadpool},
+        rpc_get_blocks{_rpc_get_blocks}
+    {
+        assert(config.pending_chunk_queue_size > 0);
+        assert(config.max_get_blocks_attempts > 0);
+    }
+
+    ~AsyncScanContextLegacy()
+    {
+        std::lock_guard<std::mutex> lock{m_async_scan_context_mutex};
+        wait_until_pending_queue_clears();
+    }
+
+    /// Kick off the scanner starting from the provided index
+    void begin_scanning_from_index(const std::uint64_t start_index, const std::uint64_t max_chunk_size_hint);
```

Comment:
```suggestion
    void begin_scanning_from_index(const std::uint64_t start_index, const std::uint64_t max_chunk_size_hint) override;
```

Make it clear this is implementing `ScanContextLedger`.